### PR TITLE
fix: Set initial progress from pod metadata if exists. Fixes #13057

### LINKED ISSUE #13057: Self reporting progress does not work when `workflows.argoproj.io/progress` is set
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I have tested with the `:latest` image tag (i.e. `quay.io/argoproj/workflow-controller:latest`) and can confirm the issue still exists on `:latest`. If not, I have explained why, **in detail**, in my description below.
- [X] I have searched existing issues and could not find a match for this bug
- [X] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/main/docs/CONTRIBUTING.md))

### What happened/what did you expect to happen?

The workflow sample in docs: https://argo-workflows.readthedocs.io/en/stable/progress/ does not work.

image

#8000 this creates a `WorkflowTaskResult` instead of doing a pod patch. However, `podReconciliation` still reconcile node progress from annotation `workflows.argoproj.io/progress`, so node progress reconciled from `WorkflowTaskResult` is overrided by the initial progress annotation on the pod: `0/100`.

### Version

latest

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: progress-
spec:
  entrypoint: main
  templates:
    - name: main
      dag:
        tasks:
          - name: progress
            template: progress
    - name: progress
      metadata:
        annotations:
          workflows.argoproj.io/progress: 0/100
      container:
        image: alpine
        imagePullPolicy: IfNotPresent
        command: [ "/bin/sh", "-c" ]
        args:
          - |
            for i in `seq 1 10`; do sleep 10; echo "$(($i*10))"'/100' > $ARGO_PROGRESS_FILE; done
```

### Logs from the workflow controller

```text
kubectl logs -n argo deploy/workflow-controller | grep ${workflow}
```

### Logs from in your workflow's wait container

```text
time="2024-05-16T03:16:59.905Z" level=info msg="Starting Workflow Executor" version=untagged
time="2024-05-16T03:16:59.910Z" level=info msg="Using executor retry strategy" Duration=1s Factor=1.6 Jitter=0.5 Steps=5
time="2024-05-16T03:16:59.911Z" level=info msg="Executor initialized" deadline="2024-05-16 03:21:58 +0000 UTC" includeScriptOutput=false namespace=argo podName=progress-ckdv7-progress-2738936019 templateName=progress version="&Version{Version:untagged,BuildDate:2024-05-09T02:07:19Z,GitCommit:6b16f24c97e4b9c792e18835304e517f2d0ddd0f,GitTag:untagged,GitTreeState:clean,GoVersion:go1.21.10,Compiler:gc,Platform:linux/amd64,}"
time="2024-05-16T03:16:59.917Z" level=debug msg="Create workflowtaskresults 201"
time="2024-05-16T03:16:59.918Z" level=info msg="Starting deadline monitor"
time="2024-05-16T03:17:06.923Z" level=debug msg="Create workflowtaskresults 409"
time="2024-05-16T03:17:06.926Z" level=debug msg="Patch workflowtaskresults 200"
time="2024-05-16T03:17:11.919Z" level=info progress=10/100
time="2024-05-16T03:17:13.923Z" level=debug msg="Create workflowtaskresults 409"
time="2024-05-16T03:17:13.927Z" level=debug msg="Patch workflowtaskresults 200"
time="2024-05-16T03:17:20.922Z" level=debug msg="Create workflowtaskresults 409"
time="2024-05-16T03:17:20.926Z" level=debug msg="Patch workflowtaskresults 200"
time="2024-05-16T03:17:20.926Z" level=info progress=20/100
time="2024-05-16T03:17:27.924Z" level=debug msg="Create workflowtaskresults 409"
time="2024-05-16T03:17:27.927Z" level=debug msg="Patch workflowtaskresults 200"
time="2024-05-16T03:17:32.919Z" level=info progress=30/100
time="2024-05-16T03:17:34.925Z" level=debug msg="Create workflowtaskresults 409"
time="2024-05-16T03:17:34.928Z" level=debug msg="Patch workflowtaskresults 200"
time="2024-05-16T03:17:41.923Z" level=debug msg="Create workflowtaskresults 409"
time="2024-05-16T03:17:41.926Z" level=debug msg="Patch workflowtaskresults 200"
time="2024-05-16T03:17:41.926Z" level=info progress=40/100
time="2024-05-16T03:17:48.923Z" level=debug msg="Create workflowtaskresults 409"
time="2024-05-16T03:17:48.927Z" level=debug msg="Patch workflowtaskresults 200"
time="2024-05-16T03:17:50.919Z" level=info progress=50/100
time="2024-05-16T03:17:55.924Z" level=debug msg="Create workflowtaskresults 409"
time="2024-05-16T03:17:55.927Z" level=debug msg="Patch workflowtaskresults 200"
time="2024-05-16T03:18:02.923Z" level=debug msg="Create workflowtaskresults 409"
time="2024-05-16T03:18:02.927Z" level=debug msg="Patch workflowtaskresults 200"
time="2024-05-16T03:18:02.927Z" level=info progress=60/100
time="2024-05-16T03:18:09.926Z" level=debug msg="Create workflowtaskresults 409"
time="2024-05-16T03:18:09.929Z" level=debug msg="Patch workflowtaskresults 200"
time="2024-05-16T03:18:11.918Z" level=info progress=70/100
time="2024-05-16T03:18:16.922Z" level=debug msg="Create workflowtaskresults 409"
time="2024-05-16T03:18:16.926Z" level=debug msg="Patch workflowtaskresults 200"
time="2024-05-16T03:18:20.919Z" level=info progress=80/100
time="2024-05-16T03:18:23.924Z" level=debug msg="Create workflowtaskresults 409"
time="2024-05-16T03:18:23.927Z" level=debug msg="Patch workflowtaskresults 200"
time="2024-05-16T03:18:30.922Z" level=debug msg="Create workflowtaskresults 409"
time="2024-05-16T03:18:30.926Z" level=debug msg="Patch workflowtaskresults 200"
time="2024-05-16T03:18:32.918Z" level=info progress=90/100
time="2024-05-16T03:18:37.923Z" level=debug msg="Create workflowtaskresults 409"
time="2024-05-16T03:18:37.927Z" level=debug msg="Patch workflowtaskresults 200"
time="2024-05-16T03:18:40.967Z" level=info msg="Main container completed" error="<nil>"
time="2024-05-16T03:18:40.967Z" level=info msg="No Script output reference in workflow. Capturing script output ignored"
time="2024-05-16T03:18:40.967Z" level=info msg="No output parameters"
time="2024-05-16T03:18:40.967Z" level=info msg="No output artifacts"
time="2024-05-16T03:18:40.967Z" level=info msg="S3 Save path: /tmp/argo/outputs/logs/main.log, key: progress-ckdv7/progress-ckdv7-progress-2738936019/main.log"
time="2024-05-16T03:18:40.967Z" level=info msg="Creating minio client using static credentials" endpoint="minio:9000"
time="2024-05-16T03:18:40.968Z" level=info msg="Saving file to s3" bucket=my-bucket endpoint="minio:9000" key=progress-ckdv7/progress-ckdv7-progress-2738936019/main.log path=/tmp/argo/outputs/logs/main.log
time="2024-05-16T03:18:40.981Z" level=info msg="Save artifact" artifactName=main-logs duration=13.623371ms error="<nil>" key=progress-ckdv7/progress-ckdv7-progress-2738936019/main.log
time="2024-05-16T03:18:40.981Z" level=info msg="not deleting local artifact" localArtPath=/tmp/argo/outputs/logs/main.log
time="2024-05-16T03:18:40.981Z" level=info msg="Successfully saved file: /tmp/argo/outputs/logs/main.log"
time="2024-05-16T03:18:40.987Z" level=debug msg="Create workflowtaskresults 409"
time="2024-05-16T03:18:40.991Z" level=debug msg="Patch workflowtaskresults 200"
time="2024-05-16T03:18:40.991Z" level=info msg="Alloc=7200 TotalAlloc=13324 Sys=20325 NumGC=5 Goroutines=10"
time="2024-05-16T03:18:40.995Z" level=debug msg="Patch workflowtaskresults 200"
```

### PR DESCRIPTION
**Pull request by jswxstw**
Fixes #13057

### Motivation

Self reporting progress does not work when `workflows.argoproj.io/progress` is set.

### Modifications

1. Set initial progress from pod metadata if exists before `podReconciliation`.
2. Prevent the node's existing progress from being overwritten with the initial progress in pod metadata: only override it when workflow uses legacy/insecure pod patch.

### Verification

Local test and UT.
Workflow demo:
```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: progress-
spec:
  entrypoint: main
  templates:
    - name: main
      dag:
        tasks:
          - name: progress
            template: progress
    - name: progress
      metadata:
        annotations:
          workflows.argoproj.io/progress: 0/100
      container:
        image: alpine
        imagePullPolicy: IfNotPresent
        command: [ "/bin/sh", "-c" ]
        args:
          - |
            for i in `seq 1 10`; do sleep 10; echo "$(($i*10))"'/100' > $ARGO_PROGRESS_FILE; done
```

### CODE REVIEW

**Comment by juliev0** [comment]
Thanks for all of the fixes @jswxstw! Any interest in joining the [Argo Workflows Sustainability Effort](https://github.com/argoproj/argo-workflows/blob/main/community/sustainability_effort.md) and geting your PRs prioritized for review?

**Comment by jswxstw** [comment]
> Thanks for all of the fixes @jswxstw! Any interest in joining the [Argo Workflows Sustainability Effort](https://github.com/argoproj/argo-workflows/blob/main/community/sustainability_effort.md) and geting your PRs prioritized for review?

@juliev0 Thank you for your invitation, but unfortunately, I need to move on to other work later.

**Comment by juliev0** [comment]
> Prevent the node's existing progress from being overwritten with the initial progress in pod metadata: only override it when workflow uses legacy/insecure pod patch.

So, does this mean that the "self reporting progress" feature would only work if the RBAC is configured to prevent writing to `WorkflowTaskResult` per [this](https://github.com/argoproj/argo-workflows/blob/b6d6403320157e614af2c14c2417433e55eadec1/workflow/executor/executor.go#L822) line?

**Comment by jswxstw** [comment]
> So, does this mean that the "self reporting progress" feature would only work if the RBAC is configured to prevent writing to `WorkflowTaskResult` per [this](https://github.com/argoproj/argo-workflows/blob/b6d6403320157e614af2c14c2417433e55eadec1/workflow/executor/executor.go#L822) line?

Not really, "self reporting progress" feature also works when `WorkflowTaskResult` is written, but setting initial progress with pod label `workflows.argoproj.io/progress: N/M` will render this feature ineffective.

**Comment by juliev0** [comment]
> 
> Not really, "self reporting progress" feature also works when `WorkflowTaskResult` is written, but setting initial progress with pod label `workflows.argoproj.io/progress: N/M` will render this feature ineffective.

Sorry. Does this mean that we're calling `ParseProgress()` somewhere else for the case of `WorkflowTaskResult` written? (since in your code I see it now only being called if `AnnotationKeyReportOutputsCompleted` is set, which seems to only get [set](https://github.com/argoproj/argo-workflows/blob/b6d6403320157e614af2c14c2417433e55eadec1/workflow/executor/executor.go#L822) if the RBAC prevents writing to `WorkflowTaskResult`)

**Comment by jswxstw** [comment]
> (since in your code I see it now only being called if `AnnotationKeyReportOutputsCompleted` is set, which seems to only get [set](https://github.com/argoproj/argo-workflows/blob/b6d6403320157e614af2c14c2417433e55eadec1/workflow/executor/executor.go#L822) if the RBAC prevents writing to `WorkflowTaskResult`)

Controller will get node progress both from `WorkflowTaskResult` and pod annotation and the latter will override the former since `podReconciliation` is executed after `taskResultReconciliation`.
Normally, progress will appear in the pod annotation only if the RBAC prevents writing to `WorkflowTaskResult`. However, users can manually set the initial progress to the pod annotation, which causes the controller always use the initial progress and ignoring the real progress in `WorkflowTaskResult`.
So, I determine if RBAC is preventing writing to `WorkflowTaskResult` by check if pod annotation `AnnotationKeyReportOutputsCompleted` is exist.

File: workflow/controller/operator.go [hunk 1]
```diff
@@ -1418,16 +1418,21 @@ func (woc *wfOperationCtx) assessNodeStatus(pod *apiv1.Pod, old *wfv1.NodeStatus
...
+			}
 		}
-		if err := json.Unmarshal([]byte(x), new.Outputs); err != nil {
-			new.Phase = wfv1.NodeError
-			new.Message = err.Error()
+
+		if x, ok = pod.Annotations[common.AnnotationKeyProgress]; ok {
+			if p, ok := wfv1.ParseProgress(x); ok {
+				new.Progress = p
+			}
```
Comment (juliev0) [inline workflow/controller/operator.go:1435]: can we add a comment above this whole code block (above line 1413) to say something like "if `AnnotationKeyReportOutputsCompleted` is set, it means RBAC prevented `WorkflowTaskResult` from being written: therefore, we use Annotations to get information from Pod"

**Comment by juliev0** [comment]
> > (since in your code I see it now only being called if `AnnotationKeyReportOutputsCompleted` is set, which seems to only get [set](https://github.com/argoproj/argo-workflows/blob/b6d6403320157e614af2c14c2417433e55eadec1/workflow/executor/executor.go#L822) if the RBAC prevents writing to `WorkflowTaskResult`)
> 
> Controller will get node progress both from `WorkflowTaskResult` and pod annotation and the latter will override the former since `podReconciliation` is executed after `taskResultReconciliation`. Normally, progress will appear in the pod annotation only if the RBAC prevents writing to `WorkflowTaskResult`. However, users can manually set the initial progress to the pod annotation, which causes the controller always use the initial progress and ignore the real progress in `WorkflowTaskResult`. So, I determine if RBAC is preventing writing to `WorkflowTaskResult` by check if pod annotation `AnnotationKeyReportOutputsCompleted` is exist.

okay, thanks for the explanation

**Comment by juliev0** [comment]
Did you test it both while having the RBAC permission and while not?

**Comment by jswxstw** [comment]
> Did you test it both while having the RBAC permission and while not? (if you can confirm that and add the comment I mentioned (or something like it), then I'll gladly approve and merge. Thanks for fixing this!

There is already a warning log above.
https://github.com/argoproj/argo-workflows/blob/16f0a8ea3cc0690d32994d309ac6b6e26a9409ca/workflow/controller/operator.go#L1425

I removed `workflowtaskresult` edit permission for executor, wait container logs shows that:
```
time="2024-08-11T03:43:30.299Z" level=debug msg="Create workflowtaskresults 403"
time="2024-08-11T03:43:30.299Z" level=warning msg="failed to patch task result, falling back to legacy/insecure pod patch, see https://argo-workflows.readthedocs.io/en/latest/workflow-rbac/" error="workflowtaskresults.argoproj.io is forbidden: User \"system:serviceaccount:argo:default\" cannot create resource \"workflowtaskresults\" in API group \"argoproj.io\" in the namespace \"argo\""
time="2024-08-11T03:43:30.308Z" level=debug msg="Patch pods 200"
time="2024-08-11T03:43:35.298Z" level=info progress=10/100
time="2024-08-11T03:43:37.300Z" level=debug msg="Create workflowtaskresults 403"
time="2024-08-11T03:43:37.300Z" level=warning msg="failed to patch task result, falling back to legacy/insecure pod patch, see https://argo-workflows.readthedocs.io/en/latest/workflow-rbac/" error="workflowtaskresults.argoproj.io is forbidden: User \"system:serviceaccount:argo:default\" cannot create resource \"workflowtaskresults\" in API group \"argoproj.io\" in the namespace \"argo\""
time="2024-08-11T03:43:37.308Z" level=debug msg="Patch pods 200"
time="2024-08-11T03:43:44.299Z" level=debug msg="Create workflowtaskresults 403"
time="2024-08-11T03:43:44.299Z" level=warning msg="failed to patch task result, falling back to legacy/insecure pod patch, see https://argo-workflows.readthedocs.io/en/latest/workflow-rbac/" error="workflowtaskresults.argoproj.io is forbidden: User \"system:serviceaccount:argo:default\" cannot create resource \"workflowtaskresults\" in API group \"argoproj.io\" in the namespace \"argo\""
time="2024-08-11T03:43:44.308Z" level=debug msg="Patch pods 200"
time="2024-08-11T03:43:44.309Z" level=info progress=20/100
```
Self reporting progress works well.

File: workflow/controller/operator.go (hunk 1 above)
Comment (jswxstw) [reply workflow/controller/operator.go:1446]: There is already a warning log above.
https://github.com/argoproj/argo-workflows/blob/16f0a8ea3cc0690d32994d309ac6b6e26a9409ca/workflow/controller/operator.go#L1425

Comment (juliev0) [reply workflow/controller/operator.go:1446]: I see, but I find that comment and the link it references to not be super clear; and moreover, the part I wanted to clarify was that you're using the fact that `pod.Annotations[common.AnnotationKeyReportOutputsCompleted]` is set to mean that we need to use Annotations to get all of the information. I don't find all of this necessarily obvious personally.

Comment (jswxstw) [reply workflow/controller/operator.go:1449]: ok. I've added some comments to clarify this.

**Review by juliev0 — APPROVED** [review]

**Pull request merged**